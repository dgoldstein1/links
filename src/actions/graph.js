import { store } from "../reducers/index";
// api
import * as kv from "../api/twowaykv";
import * as graph from "../api/biggraph";
import * as wiki from "../api/wiki";
import { _generateRoot } from "../reducers/graph";

export const SET_ROOT_NODE = "SET_ROOT_NODE";
export function setRootNode(node) {
  return {
    type: SET_ROOT_NODE,
    node
  };
}

export const SET_MAX_NEIGHBORS = "SET_MAX_NEIGHBORS";
export function setMaxNeighbors(maxNeighbors) {
  return {
    type: SET_MAX_NEIGHBORS,
    maxNeighbors
  };
}

export const SET_SELECTED_NODE_INFO = "SET_SELECTED_NODE_INFO";
export function setSelectedNodeInfo(info) {
  return {
    type: SET_SELECTED_NODE_INFO,
    ...info
  };
}

export const SET_SELECTED_NODE = "SET_SELECTED_NODE";
export function setSelectedNode(node, animate = true) {
  let selectedNodeId;
  try {
    selectedNodeId = store.getState().graph.selectedNode.node.id;
  } catch {}
  if (node.id === selectedNodeId) return;
  fetchAndStoreSelectedNodeInfo(node, animate);
  store.dispatch({
    type: SET_SELECTED_NODE,
    node
  });
}

export const SET_TARGET_NODE = "SET_TARGET_NODE";
export function setTargetNode(node) {
  return {
    type: SET_TARGET_NODE,
    node
  };
}

export const ADD_NEIGHBORS_TO_GRAPH = "ADD_NEIGHBORS_TO_GRAPH";
export function addNeighborsToGraph(node, neighbors) {
  return {
    type: ADD_NEIGHBORS_TO_GRAPH,
    node,
    neighbors
  };
}

export const SET_GRAPH_LAYOUT = "SET_GRAPH_LAYOUT";
export function setGraphLayout(newLayout) {
  return {
    type: SET_GRAPH_LAYOUT,
    layout: newLayout
  };
}

export const CLEAR_GRAPH = "CLEAR_GRAPH";
export function clearGraph() {
  return { type: CLEAR_GRAPH };
}

export const SET_GRAPH_LOADING = "SET_GRAPH_LOADING";
export function setGraphLoading(val) {
  return {
    type: SET_GRAPH_LOADING,
    newValue: val
  };
}

export const SET_GRAPH_ERROR = "SET_GRAPH_ERROR";
export function setGraphError(error) {
  return {
    type: SET_GRAPH_ERROR,
    error
  };
}

export const SET_GRAPH_PATH = "SET_GRAPH_PATH";
export function setGraphPath(path) {
  return {
    type: SET_GRAPH_PATH,
    path
  };
}

export const GRAPH_ANIMATE_TIME = 1500;
export function _animateViews() {
  store.dispatch(setGraphLayout("hierarchy"));
  setTimeout(() => {
    store.dispatch(setGraphLayout("cluster"));
  }, GRAPH_ANIMATE_TIME);
}

/**
 * fetches and store selected node information from wikipedia
 * @param node {node}
 **/
export function fetchAndStoreSelectedNodeInfo(node, animate) {
  // set as loading
  store.dispatch(setSelectedNodeInfo({ loading: true }));
  if (animate) _animateViews();
  // fetch from api
  wiki.getDescription(node.label).then(r => {
    // set result in store
    store.dispatch(
      setSelectedNodeInfo({
        ...r,
        loading: false
      })
    );
  });
}

// fetches neighbors of node and adds them to graph
export function fetchAndStoreNeighbors(
  node,
  callback = err => {},
  ignoreEmpty = false
) {
  let finalCallback = e => {
    if (e) store.dispatch(setGraphError(e));
    store.dispatch(setGraphLoading(false));
    callback(e);
  };
  store.dispatch(setGraphLoading(true));
  graph.getNeighbors(node.id).then(gr => {
    if (!gr.success) return finalCallback(gr.error);
    // don't do anything if no neighbors
    if (gr.data.length === 0 && ignoreEmpty)
      return finalCallback("no neighbors found for node '" + node.label + "'");
    // neighbor ids => values
    kv.entriesFromValues(gr.data).then(nIds => {
      if (nIds.error) return finalCallback(nIds.error);
      // success! transform data
      nIds.data.entries = nIds.data.entries || [];
      let neighbors = [];
      nIds.data.entries.forEach(n => {
        neighbors.push({
          id: n.value,
          label: n.key
        });
      });
      // set in store
      setSelectedNode(node);
      store.dispatch(addNeighborsToGraph(node, neighbors));
      return finalCallback();
    });
  });
}

// clears graph, finds neighbors, and sets new root
export function setNewRoot(node, callback = () => {}) {
  store.dispatch(setGraphLoading(true));
  store.dispatch(clearGraph());
  node = _generateRoot(node.label, node.id);
  fetchAndStoreNeighbors(node, err => {
    if (err) {
      store.dispatch(setGraphError(err));
    } else {
      store.dispatch(setRootNode(node));
    }
    store.dispatch(setGraphLoading(false));
    callback(err);
  });
}

/**
 * fetches and stores new path in store
 * @param start {node}
 * @param end {node}
 **/
export function fetchAndStorePath(start, end) {
  // inner helper function
  let _errOut = e => {
    store.dispatch(setGraphError(e));
    store.dispatch(setGraphLoading(false));
  };
  setSelectedNode(start);
  store.dispatch(setGraphLoading(true));
  graph.shortestPath(start.id, end.id).then(r => {
    if (!r.success) return _errOut(r.error);
    // path found, get entries from values
    // only get entries in middle, already have beginning and end
    kv.entriesFromValues(r.data.slice(1, -1)).then(r => {
      if (!r.success) return _errOut(r.error);
      // transfrom response to nodes
      r.data.entries = r.data.entries || [];
      let nodePath = r.data.entries.map(e => ({ id: e.value, label: e.key }));
      // insert root and end into array for sanity
      nodePath = [start, ...nodePath, end];
      // clear graph and set new path
      store.dispatch(setGraphPath(nodePath));
      store.dispatch(setGraphLoading(false));
    });
  });
}

export function setStartPath(node) {
  // target node in store, find path
  if (store.getState().graph.targetNode.id) {
    return fetchAndStorePath(node, store.getState().graph.targetNode);
  }
  // else set new root
  setNewRoot(node);
}

export function setTargetPath(node) {
  store.dispatch(setTargetNode(node));
  if (store.getState().graph.rootNode.id) {
    fetchAndStorePath(store.getState().graph.rootNode, node);
  }
}

// fetches and stores random starting node and neighbors
// callback called with string error
export function fetchAndStoreRandomStartNode(callback) {
  // fetch random node
  kv.random(1).then(r => {
    if (!r.success) return callback(r.error);
    // fetch neighbors of node
    let node = _generateRoot(r.data[0].key, r.data[0].value);
    store.dispatch(setRootNode(node));
    fetchAndStoreNeighbors(node, callback);
  });
}
