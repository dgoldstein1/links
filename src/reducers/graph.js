import * as ac from "../actions/graph";
import _ from "lodash";

const ROOT_NODE_SIZE = 10000;

const initialState = {
  rootNode: {},
  graph: {
    nodes: [],
    edges: []
  },
  loading: false,
  error: undefined,
  layout: "hierarchy"
};

const visitReducer = (state = initialState, action) => {
  switch (action.type) {
    case ac.SET_GRAPH_ERROR:
      return Object.assign({}, state, {
        error: action.error
      });
    case ac.SET_GRAPH_LAYOUT:
      return Object.assign({}, state, {
        layout: action.layout
      });
    case ac.SET_GRAPH_LOADING:
      return Object.assign({}, state, {
        loading: action.newValue
      });
    case ac.SET_ROOT_NODE:
      return Object.assign({}, state, {
        rootNode: _generateRoot(action.node.id, action.node.label)
      });
    case ac.ADD_NEIGHBORS_TO_GRAPH:
      // set initial position to source
      action.neighbors = action.neighbors.map(n => ({
        ...n,
        x: action.node.x,
        y: action.node.y,
        size: action.node.size / 2
      }));
      // get list of edges to add
      let edgesToAdd = [];
      action.neighbors.forEach(n => {
        edgesToAdd.push({
          id: `${action.node.id}->-->--${n.id}`,
          source: action.node.id,
          target: n.id
        });
      });
      // add in node to make sure it's in list of nodes
      action.neighbors.push(action.node);
      // return state
      return Object.assign({}, state, {
        graph: {
          nodes: _.unionBy(state.graph.nodes, action.neighbors, "id"),
          edges: _.unionBy(state.graph.edges, edgesToAdd, "id")
        }
      });
    default:
      return state;
  }
};

export function _generateRoot(label, id) {
  return {
    label,
    id,
    x: 0,
    y: 0,
    size: ROOT_NODE_SIZE
  };
}

export default visitReducer;
