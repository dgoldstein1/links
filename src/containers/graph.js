/* eslint-disable import/first */
import React from "react";
import { connect } from "react-redux";
import { Sigma, EdgeShapes } from "react-sigma";
import Dagre from "react-sigma/lib/Dagre";
import ForceLink from "react-sigma/lib/ForceLink";
import { fetchAndStoreNeighbors, setGraphError } from "../actions/graph";
import ErrorCard from "../components/errorCard";
import { store } from "../reducers";

class Graph extends React.Component {
  constructor() {
    super();
    this._onNodeClick = this._onNodeClick.bind(this);
    this._getGraphFromLayout = this._getGraphFromLayout.bind(this);
  }

  _onNodeClick(e) {
    // add neighbors
    fetchAndStoreNeighbors(e.data.node);
  }

  _getGraphFromLayout() {
    if (this.props.layout === "hierarchy") {
      return <Dagre directed={true} easing={true} multigraph={true} />;
    }
    // default
    return (
      <ForceLink
        slowDown={1}
        background
        easing="cubicInOut"
        randomize="globally"
        strongGravityMode={true}
        gravity={0}
      />
    );
  }

  render() {
    return (
      <div>
        {this.props.loading && !this.props.error && (
          <div className="spinner secondary" />
        )}
        {!this.props.loading && this.props.error && (
          <ErrorCard
            error={this.props.error}
            action="close"
            onAction={() => store.dispatch(setGraphError(undefined))}
            type="warning"
          />
        )}
        {!this.props.loading && !this.props.error && (
          <div>
            <Sigma
              onClickNode={this._onNodeClick}
              renderer="canvas"
              graph={this.props.graph}
              settings={{
                flex: 1,
                labelThreshold: 0,
                drawEdges: true,
                drawEdgeLabels: true,
                clone: false
              }}
              style={{ height: "700px" }}
            >
              {this._getGraphFromLayout()}
              <EdgeShapes default="tapered" />
            </Sigma>
          </div>
        )}
      </div>
    );
  }
}

// connect to store
let mapStateToProps = state => ({ ...state.graph });
export default connect(mapStateToProps)(Graph);
