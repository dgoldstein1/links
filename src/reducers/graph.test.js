import graph from "./graph";

import {
  SET_ROOT_NODE,
  ADD_NEIGHBORS_TO_GRAPH,
  SET_GRAPH_PATH
} from "../actions/graph";

describe("reducers", () => {
  describe("graph", () => {
    it("initializes with correct state", () => {
      expect(graph(undefined, { action: undefined })).toMatchSnapshot();
    });

    let initialState = {
      graph: {
        edges: [
          {
            id: "1->-->--2",
            source: 1,
            target: 2
          }
        ],
        nodes: [
          {
            id: 2,
            value: "2"
          }
        ]
      }
    };

    describe("SET_GRAPH_PATH", () => {
      let testTable = [
        {
          name: "adds normal path",
          path: [
            {
              id: 1,
              label: "1"
            },
            {
              id: 2,
              label: "2"
            },
            {
              id: 3,
              label: "3"
            }
          ],
          initialState: undefined
        },
        {
          name: "clears graph if it's already there",
          path: [
            {
              id: 1,
              label: "1"
            }
          ],
          initialState: initialState
        }
      ];

      testTable.forEach(t => {
        it(t.name, () => {
          let action = {
            type: SET_GRAPH_PATH,
            ...t
          };
          expect(graph(t.initialState, action)).toMatchSnapshot();
        });
      });
    });

    describe("ADD_NEIGHBORS_TO_GRAPH", () => {
      let testTable = [
        {
          name: "adds new values",
          node: { id: 1, value: "1" },
          neighbors: [{ id: 2, value: "2" }],
          initialState: undefined
        },
        {
          name: "does not add duplicate edges or nodes",
          node: { id: 1, value: "1" },
          initialState: initialState,
          neighbors: [{ id: 2, value: "2" }]
        },
        {
          name: "merges new values with old",
          node: { id: 1, value: "1" },
          initialState: initialState,
          neighbors: [{ id: 2, value: "2" }, { id: 3, value: "3" }]
        }
      ];

      testTable.forEach(t => {
        it(t.name, () => {
          let action = {
            type: ADD_NEIGHBORS_TO_GRAPH,
            ...t
          };
          expect(graph(t.initialState, action)).toMatchSnapshot();
        });
      });
    });
  });
});
