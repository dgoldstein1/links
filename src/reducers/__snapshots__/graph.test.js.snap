// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers graph ADD_NEIGHBORS_TO_GRAPH adds new values 1`] = `
Object {
  "error": undefined,
  "graph": Object {
    "edges": Array [
      Object {
        "id": "1->-->--2",
        "source": 1,
        "target": 2,
      },
    ],
    "nodes": Array [
      Object {
        "id": 2,
        "size": NaN,
        "value": "2",
        "x": undefined,
        "y": undefined,
      },
      Object {
        "id": 1,
        "value": "1",
      },
    ],
  },
  "layout": "hierarchy",
  "loading": false,
  "rootNode": Object {},
}
`;

exports[`reducers graph ADD_NEIGHBORS_TO_GRAPH does not add duplicate edges or nodes 1`] = `
Object {
  "graph": Object {
    "edges": Array [
      Object {
        "id": "1->-->--2",
        "source": 1,
        "target": 2,
      },
    ],
    "nodes": Array [
      Object {
        "id": 2,
        "value": "2",
      },
      Object {
        "id": 1,
        "value": "1",
      },
    ],
  },
}
`;

exports[`reducers graph ADD_NEIGHBORS_TO_GRAPH merges new values with old 1`] = `
Object {
  "graph": Object {
    "edges": Array [
      Object {
        "id": "1->-->--2",
        "source": 1,
        "target": 2,
      },
      Object {
        "id": "1->-->--3",
        "source": 1,
        "target": 3,
      },
    ],
    "nodes": Array [
      Object {
        "id": 2,
        "value": "2",
      },
      Object {
        "id": 3,
        "size": NaN,
        "value": "3",
        "x": undefined,
        "y": undefined,
      },
      Object {
        "id": 1,
        "value": "1",
      },
    ],
  },
}
`;

exports[`reducers graph initializes with correct state 1`] = `
Object {
  "error": undefined,
  "graph": Object {
    "edges": Array [],
    "nodes": Array [],
  },
  "layout": "hierarchy",
  "loading": false,
  "rootNode": Object {},
}
`;
